### PR TITLE
First schedule time is expected to be calculated using next run time

### PR DESCRIPTION
**Pull request by frsyuki**
First schedule time was calculated using next session time. But it's
expected to be calculated using next run time instead as an issue filed
at [#149].

For example, a workflow has `daily>: 10:00:00` schedule and current time is at 8am. Current behavior is that first run time becomes tomorrow 10:00:00. This is because today's "schedule time" (that is 00:00:00) is already passed.
This PR changes it to be today 10:00:00 because today's "run time" (that is 10:00:00) is not passed yet.

This PR also adds `--schedule-from` option to `digdag push` command so that test cases can be independent from the environment's current time.

### CODE REVIEW

**Comment by frsyuki** [comment]
@danielnorberg PTAL

File: digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    {
+        return Instant.from(TIME_FORMAT.parse(time));
+    }
+
+    @Test
+    public void firstScheduleTimeUtc()
+    {
+        // current time is 09:00:00
+        // schedule is 10:00:00 every day
+        // schedule at today 01:00:00
```
Comment (danielnorberg) [inline digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java]: This is a bit confusing. Do you mean 10:00:00?

File: digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        Instant currentTime1 = instant("2016-02-03 09:00:00 +0000");
```
Comment (danielnorberg) [inline digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java]: If I change this line to `Instant currentTime1 = instant("2016-02-03 01:00:00 +0000");` the test fails.

```
java.lang.AssertionError: 
Expected: is <ScheduleTime{runTime=2016-02-03T10:00:00Z, time=2016-02-03T00:00:00Z}>
     but: was <ScheduleTime{runTime=2016-02-03T01:00:00Z, time=2016-02-02T15:00:00Z}>
Expected :is <ScheduleTime{runTime=2016-02-03T10:00:00Z, time=2016-02-03T00:00:00Z}>
     
Actual   :<ScheduleTime{runTime=2016-02-03T01:00:00Z, time=2016-02-02T15:00:00Z}>
```

Is this expected?

**Comment by danielnorberg** [comment]
I feel that the scheduler code is a bit confusing. Maybe we should clean it up and/or comment it a bit.

File: digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java (hunk 1 above)
Comment (frsyuki) [inline digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java]: yes!

File: digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java (hunk 2 above)
Comment (frsyuki) [inline digdag-standards/src/test/java/io/digdag/standards/scheduler/DailySchedulerTest.java]: no. it doesn't fail on my machine. does it?
I guess that gradle on your machine is confused with some cache.

**Comment by frsyuki** [comment]
which part do you think is confusing? about having scheduled time and run time?

**Comment by danielnorberg** [comment]
Yes, I think the `time` and `runTime` fields of `ScheduleTime` is confusing. Would be nice to have some comments there to explain what they mean.

It would also be nice to have some comments explaining how `DailySchedulerFactory` is using `CronScheduler`. I.e., why it is using the pattern `0 0 * * *` and what the purpose of the `delaySeconds` argument to the `CronScheduler` constructor is.

**Comment by danielnorberg** [comment]
Interesting. The modified test fails in the JST timezone but not in the PDT timezone.

**JST**
```
➜  digdag git:(dano/first-schedule-time-fix--add-failing-test) date
Fri Jul 29 13:51:06 JST 2016
➜  digdag git:(dano/first-schedule-time-fix--add-failing-test) ./gradlew cleanTest test --info -p digdag-standards --tests '*DailySchedulerTest*'
Starting Build

<...snip...>

io.digdag.standards.scheduler.DailySchedulerTest > firstScheduleTimeUtc STARTED
Gradle Test Executor 1 finished executing tests.

io.digdag.standards.scheduler.DailySchedulerTest > firstScheduleTimeUtc FAILED
    java.lang.AssertionError:
    Expected: is <ScheduleTime{runTime=2016-02-03T10:00:00Z, time=2016-02-03T00:00:00Z}>
         but: was <ScheduleTime{runTime=2016-02-03T01:00:00Z, time=2016-02-02T15:00:00Z}>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at io.digdag.standards.scheduler.DailySchedulerTest.firstScheduleTimeUtc(DailySchedulerTest.java:57)

10 tests completed, 1 failed
Finished generating test XML results (0.013 secs) into: /Users/dano/projects/digdag/digdag-standards/build/test-results
Generating HTML test report...
Finished generating test html results (0.07 secs) into: /Users/dano/projects/digdag/digdag-standards/build/reports/tests
:digdag-standards:test FAILED
:digdag-standards:test (Thread[main,5,main]) completed. Took 3.638 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':digdag-standards:test'.
> There were failing tests. See the report at: file:///Users/dano/projects/digdag/digdag-standards/build/reports/tests/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output.

BUILD FAILED

Total time: 12.788 secs
Stopped 0 compiler daemon(s).
```

**PDT**
```
➜  digdag git:(dano/first-schedule-time-fix--add-failing-test) date
Thu Jul 28 21:50:40 PDT 2016
➜  digdag git:(dano/first-schedule-time-fix--add-failing-test) ./gradlew cleanTest test --info -p digdag-standards --tests '*DailySchedulerTest*'
Starting Build

<...snip...>

io.digdag.standards.scheduler.DailySchedulerTest > firstScheduleTimeUtc STARTED

io.digdag.standards.scheduler.DailySchedulerTest > firstScheduleTimeUtc PASSED
Gradle Test Executor 1 finished executing tests.
Finished generating test XML results (0.009 secs) into: /Users/dano/projects/digdag/digdag-standards/build/test-results
Generating HTML test report...
Finished generating test html results (0.044 secs) into: /Users/dano/projects/digdag/digdag-standards/build/reports/tests
:digdag-standards:test (Thread[main,5,main]) completed. Took 4.001 secs.

BUILD SUCCESSFUL

Total time: 12.868 secs
Stopped 0 compiler daemon(s).
```

**Comment by danielnorberg** [comment]
Reproduced on travis-ci by setting `TZ=JST`.

https://travis-ci.org/treasure-data/digdag/jobs/148224067

**Comment by frsyuki** [comment]
I found that the weird behavior is due to a bug of cron4j: https://gist.github.com/frsyuki/618c4e6c1f5f876e4ee74b9da2fd37c0
I applied a workaround fix to digdag: 86daddc0669de3b2cf7470ad4463ea75c06c7cb1.
Cron4j is not hosted on github or something similar. I sent a bug report on twitter but I'm not sure it'll be fixed or not: [1](https://twitter.com/frsyuki/status/760227626686361600) [2](https://twitter.com/frsyuki/status/760227704947912704) [3](https://twitter.com/frsyuki/status/760227823462129664)

**Comment by frsyuki** [comment]
ScheduleTime has `time` in addition to `runTime` because a session needs `session_time`.
Note: `session_time` means "for what time this session runs". Actual execution time doesn't match with `session_time` when a session attempt runs for a past time (backfill), or a session attempt starts after some delay.
So, ScheduleTime class essentially needs 2 timestamps. though we can change the names of the fields or change implementation of CronScheduler (and maybe we should).

CronScheduler calculates `time` (e.g. `2016-01-01 00:00:00` for `daily> 10:00:00`) first and adds some delay to calculate `runTime` (e.g. `2016-01-01 10:00:00` for `daily> 10:00:00`). This is simply because I imagined this implementation first. [PerfectSched](https://github.com/treasure-data/perfectsched) uses the same logic. But we don't have to keep using it.

We can change it to be opposite so that it calculates `runTime` first, and then truncate it to calculate `time`. With this design, CronSchedulerFactory will have `_command: PATTERN` and `session_time_truncation: ENUM` options, where it has now `_command: PATTERN` and `delay: SECONDS` options. The exceptional case of daylight saving times will not happen. Although there should be a discussion about what's the best behavior.

This discussion should be done by another PR.

**Comment by danielnorberg** [comment]
Interesting. So `session_time` for `daily> HH:MM:SS` is always `00:00:00`? I did not know that, but I guess it makes sense. Hopefully our users have a better intuition about that than me.

The docs do not seem to mention what `session_time` the different `schedule:` directives result in.

http://www.digdag.io/scheduling_workflow.html

**Comment by danielnorberg** [comment]
👍

**Comment by frsyuki** [comment]
> So session_time for daily> HH:MM:SS is always 00:00:00?

Yes.

**Pull request merged**